qoffer-infrastructure/src/main/groovy/life/qbic/portal/qoffer2/customers/CustomerDbConnector.groovy

```diff
@@ -81,23 +152,122 @@ class CustomerDbConnector implements CreateCustomerDataSource, UpdateCustomerDat
   @Override
   void addCustomer(Customer customer) throws DatabaseQueryException {
     try {
-      databaseQueries.addCustomer(customer)
-    } catch (DatabaseQueryException e) {
-      throw new DatabaseQueryException(e.message)
+      if (customerExists(customer)) {
+        throw new DatabaseQueryException("Customer is already in the database.")
+      }
+      Connection connection = connectionProvider.connect()
+      connection.setAutoCommit(false)
+
+      connection.withCloseable {it ->
+        try {
+          int customerId = createNewCustomer(it, customer)
+          storeAffiliation(it, customerId, customer.affiliations)
+          connection.commit()
+        } catch (Exception e) {
+          log.error(e.message)
+          log.error(e.stackTrace.join("\n"))
+          connection.rollback()
+          connection.close()
+          throw new DatabaseQueryException("Could not create customer.")
+        }
+      }
+    } catch (DatabaseQueryException ignored) {
+      throw new DatabaseQueryException("The customer could not be created: ${customer.toString()}")
     } catch (Exception e) {
       log.error(e)
       log.error(e.stackTrace.join("\n"))
       throw new DatabaseQueryException("The customer could not be created: ${customer.toString()}")
     }
   }
+  private boolean customerExists(Customer customer) {
+    String query = "SELECT * FROM customer WHERE first_name = ? AND last_name = ? AND email = ?"
+    Connection connection = connectionProvider.connect()
+
+    def customerAlreadyInDb = false
+
+    connection.withCloseable {
+      def statement = connection.prepareStatement(query)
+      statement.setString(1, customer.firstName)
+      statement.setString(2, customer.lastName)
+      statement.setString(3, customer.eMailAddress)
+      statement.execute()
+      def result = statement.getResultSet()
+      customerAlreadyInDb = result.next()
+    }
+    return customerAlreadyInDb
+  }
+  private static int createNewCustomer(Connection connection, Customer customer) {
+    String query = "INSERT INTO customer (first_name, last_name, title, email) " +
+            "VALUES(?, ?, ?, ?)"
+
+    List<Integer> generatedKeys = []
+
+    def statement = connection.prepareStatement(query, Statement.RETURN_GENERATED_KEYS)
+    statement.setString(1, customer.firstName )
+    statement.setString(2, customer.lastName)
+    statement.setString(3, customer.title.value)
+    statement.setString(4, customer.eMailAddress )
+    statement.execute()
+    def keys = statement.getGeneratedKeys()
+    while (keys.next()){
+      generatedKeys.add(keys.getInt(1))
+    }
+
+    return generatedKeys[0]
```

Comment:
we should check here, that only one key has been generated

Reply:
I will not address this in this PR since this PR only moves existing code into the DBConnector.

Reply:
Is this reported as an issue or so to keep track of this?